posthog/models/cohort/dependencies.py

```diff
@@ -0,0 +1,157 @@
+    # Check if already warm
+    if cache.get(team_cache_key):
+        return
+
+    logger.info(f"Warming cohort dependency cache for team {team_id}")
+
+    try:
+        cohorts = Cohort.objects.filter(team_id=team_id, deleted=False)
+        for cohort in cohorts:
+            dependencies = extract_cohort_dependencies(cohort)
+
+            cache.set(f"cohort:deps:{cohort.id}", list(dependencies), timeout=DEPENDENCY_CACHE_TIMEOUT)
+
+            # Update reverse dependency indexes
+            for dep_id in dependencies:
+                refs_key = f"cohort:refs:{dep_id}"
+                refs = cache.get(refs_key, [])
+                if cohort.id not in refs:
+                    refs.append(cohort.id)
+                cache.set(refs_key, refs, timeout=DEPENDENCY_CACHE_TIMEOUT)
+
+        # Mark team as cached
+        cache.set(team_cache_key, True, timeout=TEAM_CACHE_TIMEOUT)
+
+        logger.info(f"Completed warming cohort dependency cache for team {team_id}, processed {len(cohorts)} cohorts")
+
+    except Exception as e:
+        logger.exception(f"Failed to warm cohort dependency cache for team {team_id}", error=e)
+        raise
+
+
+def _clear_team_cache(team_id: int, cohorts) -> None:
+    """Clear existing cache entries for a team's cohorts."""
+    # Just clear the team cache marker. The other keys will be overwritten the next time the cache is cleared.
+    cache.delete(f"team:deps_cached:{team_id}")
+
+
+@receiver(post_save, sender="posthog.Cohort")
+def update_cohort_dependency_cache(sender, instance, **kwargs):
+    """
+    Updates existing cohort dependency cache entries based on the current state of the cohort.
+    """
+    team_cache_key = f"team:deps_cached:{instance.team_id}"
+
+    # Only update if team cache is already warmed
+    if not cache.get(team_cache_key):
+        logger.debug(f"Skipping cache update for cohort {instance.id} - team cache not warmed")
+        return
+
+    try:
+        old_deps = set(cache.get(f"cohort:deps:{instance.id}", []))
+        new_deps = extract_cohort_dependencies(instance)
+
+        # Update forward cache
+        cache.set(f"cohort:deps:{instance.id}", list(new_deps), timeout=DEPENDENCY_CACHE_TIMEOUT)
+
+        # Update reverse indexes for removed dependencies
+        for dep_id in old_deps - new_deps:
+            refs_key = f"cohort:refs:{dep_id}"
+            refs = cache.get(refs_key, [])
+            if instance.id in refs:
+                refs.remove(instance.id)
+                cache.set(refs_key, refs, timeout=DEPENDENCY_CACHE_TIMEOUT)
+
+        # Update reverse indexes for added dependencies
+        for dep_id in new_deps - old_deps:
+            refs_key = f"cohort:refs:{dep_id}"
+            refs = cache.get(refs_key, [])
+            if instance.id not in refs:
+                refs.append(instance.id)
+                cache.set(refs_key, refs, timeout=DEPENDENCY_CACHE_TIMEOUT)
+
+        logger.debug(f"Updated dependency cache for cohort {instance.id}")
```

Comment:
Make sure to run all this logic in `transaction.on_commit` like we do in `posthog/models/feature_flag/local_evaluation.py`.

One easy way to do this is to add an inner function that does the actual work and then call `transaction.on_commit(lambda: inner_function(...))`